File: edukate-storage/src/main/java/io/github/sanyavertolet/edukate/storage/Storage.java
package io.github.sanyavertolet.edukate.storage;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.nio.ByteBuffer;
import java.util.Collection;

public interface Storage<Key, Metadata> extends ReadOnlyStorage<Key, Metadata> {

    default Mono<Key> upload(Key key, String contentType, Flux<ByteBuffer> content) {
        return content.collectList().flatMap(buffers -> {
            int totalSize = buffers.stream().mapToInt(ByteBuffer::remaining).sum();
            return upload(key, totalSize, contentType, Flux.fromIterable(buffers));
        });
    }

    Mono<Key> upload(Key key, long contentLength, String contentType, Flux<ByteBuffer> content);

    Mono<Boolean> move(Key source, Key target);

    Mono<Boolean> delete(Key key);

    @SuppressWarnings("unused")
    Mono<Boolean> deleteAll(Collection<Key> keys);
}

File: edukate-storage/src/main/kotlin/io/github/sanyavertolet/edukate/storage/Storage.kt
package io.github.sanyavertolet.edukate.storage

import java.nio.ByteBuffer
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

interface Storage<Key : Any, Metadata> : ReadOnlyStorage<Key, Metadata> {
    fun upload(key: Key, contentType: String, content: Flux<ByteBuffer>): Mono<Key> =
        content.collectList().flatMap { buffers ->
            val totalSize = buffers.sumOf { it.remaining() }.toLong()
            upload(key, totalSize, contentType, Flux.fromIterable(buffers))
        }

    fun upload(key: Key, contentLength: Long, contentType: String, content: Flux<ByteBuffer>): Mono<Key>

    fun move(source: Key, target: Key): Mono<Boolean>

    fun delete(key: Key): Mono<Boolean>

    fun deleteAll(keys: Collection<Key>): Mono<Boolean>
}
